sentry-ruby/lib/sentry/metrics/aggregator.rb

```diff
@@ -0,0 +1,185 @@
+# frozen_string_literal: true
+
+module Sentry
+  module Metrics
+    class Aggregator
+      include LoggingHelper
+
+      FLUSH_INTERVAL = 5
+      ROLLUP_IN_SECONDS = 10
+
+      KEY_SANITIZATION_REGEX = /[^a-zA-Z0-9_\/.-]+/
+      VALUE_SANITIZATION_REGEX = /[^[[:word:]][[:digit:]][[:space:]]_:\/@\.{}\[\]$-]+/
+
+      METRIC_TYPES = {
+        c: CounterMetric,
+        d: DistributionMetric,
+        g: GaugeMetric,
+        s: SetMetric
+      }
+
+      # exposed only for testing
+      attr_reader :thread, :buckets, :flush_shift
+
+      def initialize(configuration, client)
+        @client = client
+        @logger = configuration.logger
+
+        @default_tags = {}
+        @default_tags['release'] = configuration.release if configuration.release
+        @default_tags['environment'] = configuration.environment if configuration.environment
+
+        @thread = nil
+        @exited = false
+        @mutex = Mutex.new
+
+        # buckets are a nested hash of timestamp -> bucket keys -> Metric instance
+        @buckets = {}
+
+        # the flush interval needs to be shifted once per startup to create jittering
+        @flush_shift = Random.rand * ROLLUP_IN_SECONDS
+      end
+
+      def add(type,
+              key,
+              value,
+              unit: 'none',
+              tags: {},
+              timestamp: nil)
+        return unless ensure_thread
+        return unless METRIC_TYPES.keys.include?(type)
+
+        timestamp = timestamp.to_i if timestamp.is_a?(Time)
+        timestamp ||= Sentry.utc_now.to_i
+
+        # this is integer division and thus takes the floor of the division
+        # and buckets into 10 second intervals
+        bucket_timestamp = (timestamp / ROLLUP_IN_SECONDS) * ROLLUP_IN_SECONDS
+
+        serialized_tags = serialize_tags(get_updated_tags(tags))
+        bucket_key = [type, key, unit, serialized_tags]
+
+        @mutex.synchronize do
+          @buckets[bucket_timestamp] ||= {}
+
+          if @buckets[bucket_timestamp][bucket_key]
+            @buckets[bucket_timestamp][bucket_key].add(value)
+          else
+            @buckets[bucket_timestamp][bucket_key] = METRIC_TYPES[type].new(value)
+          end
+        end
+      end
```

Comment:
I guess considering a force flush when the new weight exceeds the limit follows in another PR?

Reply:
yes